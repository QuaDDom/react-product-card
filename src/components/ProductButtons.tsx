import React from 'react'
import { CSSProperties, useCallback, useContext  } from 'react'
import { ProductContext } from '../context/ProductContext';
import styles from '../styles/styles.module.css'

interface ProductsI{
    className?: string,
    style?: CSSProperties,
}

export default function ProductButtons({ className, style }: ProductsI) {
    const { counter, handleClick, maxCount } = useContext(ProductContext);

    const isMaxReached = useCallback(
        () => !!maxCount && counter === maxCount,
        [counter, maxCount],
    )

    return (
    <div className={`${styles.buttonsContainer} ${className}`} style={style || {}} >
        <button className={styles.buttonMinus} onClick={()=>handleClick(-1)}>-</button>
        <div className={styles.countLabel}>{counter}</div>
        <button className={`${styles.buttonAdd} ${isMaxReached() && styles.disabled}`} onClick={()=>handleClick(+1)}>+</button>
    </div>
    )
}