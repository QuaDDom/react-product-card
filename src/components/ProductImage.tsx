import React from 'react'
import { CSSProperties, useContext } from 'react';
import styles from '../styles/styles.module.css';
import noImage from '../assets/no-image.jpg';
import { ProductContext } from '../context/ProductContext';

interface ProductImagePropsI{
    img?: string,
    style?: CSSProperties,
    className?: string
}

export default function ProductImage({img='', style, className}: ProductImagePropsI) {
    const { product } = useContext(ProductContext);
    let imgToShow: string;

    if(img){
        imgToShow = img;
    } else if(product.image){
        imgToShow = product.image;
    } else{
        imgToShow = noImage;
    }
   

    return (
        <img className={`${className} ${styles.productImg}`} src={imgToShow} style={ style } alt="Product" />
    )
}
