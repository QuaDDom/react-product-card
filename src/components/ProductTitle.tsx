import React from 'react'
import { useContext } from 'react';
import { ProductContext } from '../context/ProductContext';

interface PropsI{
    title?: string,
    style?: string,
    className?: string
}

export default function ProductTitle({title='', style, className}: PropsI) {
    const { product } = useContext(ProductContext);
    return (
        <p className={`${style} ${className}`}>{title ? title : product.title}</p>

    )
}

