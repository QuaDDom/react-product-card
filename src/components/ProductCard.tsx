import React from 'react'
import { CSSProperties } from 'react';
import styles from '../styles/styles.module.css';
import ProductTitle from './ProductTitle';
import ProductImage from './ProductImage';
import ProductButtons from './ProductButtons';
import noImage from '../assets/no-image.jpg';
import { ProductContext } from '../context/ProductContext';
import { useProduct } from '../hooks/useProduct';
import { InitialValuesI, OnChangeArgsI, ProductCardHandlersI } from '../interfaces/interfaces';


interface PropsI {
    // children?: ReactElement | ReactElement[],
    children: ( args: ProductCardHandlersI )=> JSX.Element,
    product?: ProductI,
    className?: string,
    style?: CSSProperties,
    onChange?: (args: OnChangeArgsI)=> void,
    value?: number,
    initialValues?: InitialValuesI
}

interface ProductI{
    id: string,
    title: string,
    image?: string,
    count?: number
}

const productDefault = {
    id: '1',
    title: 'No Title',
    image: noImage
}

export default function ProductCard({ children, product = productDefault, className, style, value, initialValues }:PropsI) {
    const { Provider } = ProductContext;
    const {counter, handleClick, maxCount, isMaxCountReached, reset} = useProduct({
        product,
        value,
        initialValues
    });

    return ( 
        <Provider value={{
            counter,
            handleClick,
            maxCount,
            product
        }}>
            <div  className={`${styles.productCard} ${className}`} style={style || {}}>
                { 
                    children({
                        count: counter,
                        isMaxCountReached,
                        maxCount,
                        product,
                        handleClick,
                        reset
                    }) 
                }
            </div>
        </Provider>
    )
}

ProductCard.Title = ProductTitle;
ProductCard.Image = ProductImage;
ProductCard.Buttons = ProductButtons;