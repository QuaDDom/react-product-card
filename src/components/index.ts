import  ProductCard from './ProductCard';
import  ProductButtons  from './ProductButtons';
import  ProductImage  from './ProductImage';
import  ProductTitle  from './ProductTitle';

export {ProductCard, ProductButtons, ProductImage, ProductTitle }


export const ProductCardObj = Object.assign( ProductCard, {
    Title: ProductTitle,
    Image: ProductImage,
    Buttons: ProductButtons
})


export default ProductCard;