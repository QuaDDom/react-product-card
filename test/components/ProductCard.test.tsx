import React from 'react'
import renderer from 'react-test-renderer'
import { ProductCard } from '../../src/components'
import { product } from '../data/products'

const { act } = renderer 

describe('ProductCard', ()=>{
    test('Show component correctly', ()=>{
        const wrapper = renderer.create(
            <ProductCard product={product}>
                {
                    ()=>(
                        <p>Test</p>
                    )
                }
            </ProductCard>
        )
        expect( wrapper.toJSON() ).toMatchSnapshot();
    })
    test('Increase counter', ()=>{
        const wrapper = renderer.create(
            <ProductCard>
                {
                    ({ count, handleClick })=>(
                        <>
                        <button onClick={()=> handleClick(-1)}>-</button>
                        <p>{ count }</p>
                        <button onClick={()=> handleClick(+1)}>+</button>
                        </>
                    )
                }
            </ProductCard>
        )
        let tree = wrapper.toJSON();
        expect( tree ).toMatchSnapshot();

        act(()=>(tree as any).children[2].props.onClick())

        tree = wrapper.toJSON();

        expect( (tree as any).children[1].children[0] ).toBe('1');

    })
})