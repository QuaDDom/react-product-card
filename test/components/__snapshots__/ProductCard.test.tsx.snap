// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard Increase counter 1`] = `
<div
  className="productCard undefined"
  style={Object {}}
>
  <button
    onClick={[Function]}
  >
    -
  </button>
  <p>
    0
  </p>
  <button
    onClick={[Function]}
  >
    +
  </button>
</div>
`;

exports[`ProductCard Show component correctly 1`] = `
<div
  className="productCard undefined"
  style={Object {}}
>
  <p>
    Test
  </p>
</div>
`;
